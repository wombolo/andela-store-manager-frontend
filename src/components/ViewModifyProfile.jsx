import React, { Component } from 'react';
import {
  getSingleProfile, editProfile
} from '../actions/profileActions';
import PropTypes from "prop-types";
import {connect} from "react-redux";
import Notify from "../utils/Notify";

export class ViewModifyProfile extends Component {
  state= {
    id: 1, firstname: '', lastname: '',
    email: '', role: 'admin', image: 'team-7.jpg',
  };

  componentDidMount() {
    const { id } = this.props.match.params;
    this.props.getSingleProfile(id);
  }

  componentWillReceiveProps(nextProps) {
    if(nextProps.profile) {
      const {profile} = nextProps;
      this.setState(profile);
    }
  }

  handleSubmit = (event) => {
    event.preventDefault();
    if (this.state.password === this.state.conf_password){
      this.props.editProfile(this.state);
    }
    else {
      Notify.notifyError('Passwords Mismatch.');
    }
  };

  handleChange = (event) => {
    this.setState({[event.target.id]: event.target.value});
  };


  render() {
    const { profile, authProfile } = this.props;

    // const imageFile = require(`../assets/images/profile/${this.state.image}`);

    return (
      <div className="row clearfix products-grid">
        <div className="col-10 m-0-auto m-t-5">
          <h3>{profile.firstname} {profile.lastname}'s Profile</h3>

          <form onSubmit={this.handleSubmit}>

            <div className="row">

              <div className="col-3">
                <figure className="image-holder">
                  <img alt={'image'} src={require(`../assets/images/profile/${this.state.image}`)} />
                  <input type={'file'} id={'image_file'}
                         onChange={ (e) => this.handleChange({
                           target:{
                             id:'image_file',
                             value: e.target.files[0],
                           }
                         })
                         }/>
                </figure>
              </div>


              <div className="col-7">
                <table border="0" width="100%" className="product-single-table table-striped">
                  <tbody>
                  <tr>
                    <td><label htmlFor="title">First Name</label></td>
                    <td><input id="firstname" type="text" required={true} defaultValue={this.state.firstname} onChange={this.handleChange}/></td>
                  </tr>

                  <tr>
                    <td><label htmlFor="title">Last Name</label></td>
                    <td><input id="lastname" type="text" required={true} defaultValue={this.state.lastname} onChange={this.handleChange}/></td>
                  </tr>

                  <tr>
                    <td><label htmlFor="title">Email</label></td>
                    <td>{this.state.email}</td>
                  </tr>

                  {authProfile.role === 'admin' ?
                    <tr>
                      <td><label htmlFor="title">Role</label></td>
                      <td>
                        <select id={'role'} defaultValue={this.state.role} onChange={this.handleChange}>
                          <option value='admin'>Admin</option>
                          <option value='store_attendant'>Attendant</option>
                        </select>
                      </td>
                    </tr> :
                    <tr>
                      <td><label htmlFor="title">Role</label></td>
                      <td>{this.state.role}</td>
                    </tr>
                  }

                  {authProfile.id === this.state.id ?
                    <React.Fragment>
                      <tr>
                        <td><label htmlFor="title">Password</label></td>
                        <td><input id="password" type="password" onChange={this.handleChange}/></td>
                      </tr>
                      <tr>
                        <td><label htmlFor="title">Confirm Password</label></td>
                        <td><input id="conf_password" type="password" onChange={this.handleChange}/></td>
                      </tr>
                    </React.Fragment>
                    : null
                  }
                  <tr>
                    <td>Total number of sale records created</td>
                    <td><span className="item-price">67</span></td>
                  </tr>


                  <tr>
                    <td>Total number of products sold</td>
                    <td><span>125</span></td>
                  </tr>

                  <tr>
                    <td>Total worth of goods sold</td>
                    <td><span>$9535</span></td>
                  </tr>

                  <tr>
                    <td>&nbsp;</td>
                    <td><button type="submit" className="btn btn-primary">Update</button></td>
                  </tr>
                  </tbody>

                </table>
              </div>
            </div>
          </form>
        </div>
      </div>
    )
  }
}

ViewModifyProfile.propTypes = {
  getSingleProfile: PropTypes.func.isRequired,
  editProfile: PropTypes.func.isRequired,
};


export const mapStateToProps = state => ({
  profile: state.profile.profile,
  authProfile: state.auth.user.profile,
});

export default connect(mapStateToProps, {
  getSingleProfile, editProfile
})(ViewModifyProfile);
