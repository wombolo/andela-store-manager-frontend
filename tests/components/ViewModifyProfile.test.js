import 'babel-polyfill'
import React from 'react';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';
import { MemoryRouter } from 'react-router-dom';

import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';

import { ViewModifyProfile } from '../../src/components/ViewModifyProfile.jsx';

global.confirm = () => true;


let wrapper;

const props = {
  match:{ params: 1},
  getSingleProfile: () => (
    {
      product: {id: 2, title: "Bonjour", image: "products/product-5.png",price: 150, quantity: 3, ViewModifyProfileQty: '3', status: "active"}}
    ),

  editProfile: () => Promise.resolve(),

  authProfile: {
    id: 1, firstname: 'Akano', lastname: 'Peter',
    email: 'elkenzom1@gmail.com', role: 'admin', image: 'team-7.jpg',
  },

  profile: {
    id: 1, firstname: 'Akano', lastname: 'Peter',
    email: 'elkenzom1@gmail.com', role: 'admin', image: 'team-7.jpg',
  },
};


const mockStore = configureStore([thunk]);

const store = mockStore({ ...props });


describe('ViewModifyProfile Component', () => {
  it('should match snapshot', () => {
    wrapper = shallow(<ViewModifyProfile {...props}/>);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Methods in Class', () => {
    beforeEach(() => {
      wrapper = mount(
        <Provider store={store}>
          <MemoryRouter>
          <ViewModifyProfile {...props} />
          </MemoryRouter>
        </Provider>,
    );
    });

    afterEach(() => {
      wrapper.unmount();
    });


    it('triggers handleSubmit() in ViewModifyProfileComp', () => {
      const ViewModifyProfileComp = wrapper.find('ViewModifyProfile');
      sinon.spy(ViewModifyProfileComp.instance(), 'handleSubmit');

      ViewModifyProfileComp.instance().handleSubmit({preventDefault: () => {}});
      expect(ViewModifyProfileComp.instance().handleSubmit.calledOnce)
        .toEqual(true);
    });


    it('triggers handleChange() for input title', () => {
      const wrapperNew = shallow(<ViewModifyProfile {...props} />);
      wrapperNew.setProps({product: props.product});

      const ViewModifyProfileComp = wrapper.find('ViewModifyProfile');

      const file = {lastModified: 1547454659333,
        lastModifiedDate: 'Mon Jan 14 2019 09:30:59 GMT+0100 (West Africa Standard Time) {}',
        name: "15158524_xl.jpg",
        size: 3734559,
        type: "image/jpeg",
        webkitRelativePath: ""};

      const image_file = ViewModifyProfileComp.find('#image_file').first();
      image_file.simulate('change', { target: { files: [file] , id: 'image_file'} });

      sinon.spy(ViewModifyProfileComp.instance(), 'handleChange');
      ViewModifyProfileComp.instance().handleChange({target: { value: 'u' , id: 'title'} });

      expect(ViewModifyProfileComp.instance().handleChange.calledOnce)
        .toEqual(true);
    });

  });
});
