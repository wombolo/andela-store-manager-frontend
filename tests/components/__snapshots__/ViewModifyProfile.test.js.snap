// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewModifyProfile Component should match snapshot 1`] = `
<div
  className="row clearfix products-grid"
>
  <div
    className="col-10 m-0-auto m-t-5"
  >
    <h3>
      Akano
       
      Peter
      's Profile
    </h3>
    <form
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <div
          className="col-3"
        >
          <figure
            className="image-holder"
          >
            <img
              alt="image"
              src=""
            />
            <input
              id="image_file"
              onChange={[Function]}
              type="file"
            />
          </figure>
        </div>
        <div
          className="col-7"
        >
          <table
            border="0"
            className="product-single-table table-striped"
            width="100%"
          >
            <tbody>
              <tr>
                <td>
                  <label
                    htmlFor="title"
                  >
                    First Name
                  </label>
                </td>
                <td>
                  <input
                    defaultValue=""
                    id="firstname"
                    onChange={[Function]}
                    required={true}
                    type="text"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <label
                    htmlFor="title"
                  >
                    Last Name
                  </label>
                </td>
                <td>
                  <input
                    defaultValue=""
                    id="lastname"
                    onChange={[Function]}
                    required={true}
                    type="text"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <label
                    htmlFor="title"
                  >
                    Email
                  </label>
                </td>
                <td />
              </tr>
              <tr>
                <td>
                  <label
                    htmlFor="title"
                  >
                    Role
                  </label>
                </td>
                <td>
                  <select
                    defaultValue="admin"
                    id="role"
                    onChange={[Function]}
                  >
                    <option
                      value="admin"
                    >
                      Admin
                    </option>
                    <option
                      value="store_attendant"
                    >
                      Attendant
                    </option>
                  </select>
                </td>
              </tr>
              <tr>
                <td>
                  <label
                    htmlFor="title"
                  >
                    Password
                  </label>
                </td>
                <td>
                  <input
                    id="password"
                    onChange={[Function]}
                    type="password"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <label
                    htmlFor="title"
                  >
                    Confirm Password
                  </label>
                </td>
                <td>
                  <input
                    id="conf_password"
                    onChange={[Function]}
                    type="password"
                  />
                </td>
              </tr>
              <tr>
                <td>
                  Total number of sale records created
                </td>
                <td>
                  <span
                    className="item-price"
                  >
                    67
                  </span>
                </td>
              </tr>
              <tr>
                <td>
                  Total number of products sold
                </td>
                <td>
                  <span>
                    125
                  </span>
                </td>
              </tr>
              <tr>
                <td>
                  Total worth of goods sold
                </td>
                <td>
                  <span>
                    $9535
                  </span>
                </td>
              </tr>
              <tr>
                <td>
                  Â 
                </td>
                <td>
                  <button
                    className="btn btn-primary"
                    type="submit"
                  >
                    Update
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </form>
  </div>
</div>
`;
